jax/_src/random.py

```diff
@@ -44,308 +45,102 @@
+KeyArray = Union[Array, prng.PRNGKeyArray]
 
-_UINT_DTYPES = {8: jnp.uint8, 16: jnp.uint16, 32: jnp.uint32, 64: jnp.uint64}
-
-
-def PRNGKey(seed: int) -> jnp.ndarray:
-  """Create a pseudo-random number generator (PRNG) key given an integer seed.
-
-  Args:
-    seed: a 64- or 32-bit integer used as the value of the key.
-
-  Returns:
-    A PRNG key, which is modeled as an array of shape (2,) and dtype uint32. The
-    key is constructed from a 64-bit seed by effectively bit-casting to a pair
-    of uint32 values (or from a 32-bit seed by first padding out with zeros).
-  """
-  # Avoid overflowerror in X32 mode by first converting ints to int64.
-  # This breaks JIT invariance of PRNGKey for large ints, but supports the
-  # common use-case of instantiating PRNGKey with Python hashes in X32 mode.
-  if isinstance(seed, int):
-    seed_arr = jnp.asarray(np.int64(seed))
-  else:
-    seed_arr = jnp.asarray(seed)
-  if seed_arr.shape:
-    raise TypeError(f"PRNGKey seed must be a scalar; got {seed!r}.")
-  if not np.issubdtype(seed_arr.dtype, np.integer):
-    raise TypeError(f"PRNGKey seed must be an integer; got {seed!r}")
-
-  convert = lambda k: lax.reshape(lax.convert_element_type(k, np.uint32), [1])
-  k1 = convert(lax.shift_right_logical(seed_arr, lax._const(seed_arr, 32)))
-  k2 = convert(jnp.bitwise_and(seed_arr, np.uint32(0xFFFFFFFF)))
-  return lax.concatenate([k1, k2], 0)
-
-def _is_prng_key(key: jnp.ndarray) -> bool:
-  try:
-    return key.shape == (2,) and key.dtype == np.uint32
-  except AttributeError:
-    return False
+UINT_DTYPES = prng.UINT_DTYPES
 
 
 ### utilities
 
 
-def _make_rotate_left(dtype):
-  if not jnp.issubdtype(dtype, np.integer):
-    raise TypeError("_rotate_left only accepts integer dtypes.")
-  nbits = np.array(jnp.iinfo(dtype).bits, dtype)
-
-  def _rotate_left(x, d):
-    if lax.dtype(d) != dtype:
-      d = lax.convert_element_type(d, dtype)
-    if lax.dtype(x) != dtype:
-      x = lax.convert_element_type(x, dtype)
-    return lax.shift_left(x, d) | lax.shift_right_logical(x, nbits - d)
-  return _rotate_left
-
-
-def _bit_stats(bits):
-  """This is a debugging function to compute the statistics of bit fields."""
-  return np.array([list(map(int, np.binary_repr(x, 64))) for x in bits]).mean(0)
-
-
-### hash function and split
+# TODO(mattjj,jakevdp): add more info to error message, use this utility more
+def _asarray(x):
+  """A more restrictive jnp.asarray, only accepts JAX arrays and np.ndarrays."""
+  if not isinstance(x, (np.ndarray, jnp.ndarray)):
+    raise TypeError(f"Function requires array input, got {x} of type {type(x)}.")
+  return jnp.asarray(x)
+
+def _check_prng_key(key):
+  # TODO(frostig): remove once we always enable_custom_prng
+  if type(key) is prng.PRNGKeyArray:
+    return key, False
+  elif _arraylike(key):
+    if config.jax_enable_custom_prng:
+      warnings.warn(
+          'Raw arrays as random keys to jax.random functions are deprecated. '
```

Comment:
Will this warning not trigger for every user if the default is `custom_prng=False`? Maybe there should be a hint on how to resolve the warning if that is the case?

More generally, do we expect users to turn on this `custom_prng` flag themselves? As I understand it, it's only a small group of users who rely on the keys being `jnp` arrays, and for most users this will be a no-op change, so I'm not sure they should be aware of the flag. I might be wrong though!

Reply:
The `config.jax_enable_custom_prng` flag defaults to false and the warning is guarded by it, so that (I think) it should only trigger when the flag is explicitly enabled.

In general, I tried to ensure that no new warnings would appear when the flag is at its default value (meaning no upgrade). I checked this visually by running `random_test.py` and `api_test.py`.

I don't expect end users to turn this flag on. I hope to work with some neighboring libraries to upgrade by turning it on, seeing some warnings (rather than errors), and making changes to handle them. Then we can deprecate the flag altogether and remove all these checks and warnings, though I will probably flip the default value of the flag to `True` for a while first beforehand.

Does that seem right?

Reply:
Seems good, thank you for trying to do this smoothly.

Speaking for NetKet, however, we enforce [semver](https://semver.org) on our dependencies so if you were to follow it too you could also avoid doing all this. 

We are upper bounding the minor version of `jax` to `0.2`, so if you were to interpret this change as a breaking change (which IMO it is) and bump the next version of jax to `0.3` instead of `0.2.X`,  pip would avoid installing jax 0.3 with current versions of netket, and we would have the time to upgrade our code and release a new version that is compatible with `jax 0.3`.

I don't know if jax is following semver or not, however. 
It really does make things easier for dependent packages.